doc/development_process/api_lifecycle.rst

```diff
@@ -77,16 +77,70 @@ An API can be declared ``stable`` after fulfilling the following requirements:
 - Stable APIs can get backward compatible updates, bug fixes and security fixes
   at any time.
 
+Introducing incompatible changes
+================================
+
+A stable API, as described above strives to remain backwards-compatible through
+its life-cycle. There are however cases where fulfilling this objective prevents
+technical progress or is simply unfeasible without unreasonable burden on the
+maintenance of the API and its implementation(s).
+
+In order to restrict and control the introduction of a change that breaks the
+promise of backwards compatibility the following steps must be followed whenever
+such a change is considered necessary in order to accept it in the project:
+
+#. An RFC issue must be opened on GitHub with the following content:
+
+   .. code-block:: none
+
+      Title:     RFC: API Change: <subsystem>
+      Contents:  - Background information
+                 - Details on the proposed change
+                   - Function call changes
+                   - Device Tree changes (source and bindings)
+                   - Kconfig option changes
+                 - Impact to users of the API, including the steps required
+                   to adapt out-of-tree users of the API to the change
+
+#. A Pull Request must be opened on GitHub that includes the following:
```

Comment:
I'm confused from the way this is written if the PR has to be opened simultaneously to the RFC. I would expect the RFC to be discussed and agreed upon before the PR is brought out of draft state, as changing all in-tree users is a lot of work.

Reply:
I am confused as well. Why the requirement for a separate RFC issue _and_ a PR? In my mind, a PR is just an issue with code attached, and since API changes are code changes I think (at least in some cases) they may be better documented/presented as actual code changes as opposed to text in a GitHub issue.

I suggest rewording to allow both an RFC issue and a PR or just a PR.

Reply:
I would not ask for a PR with an RFC, first RFC, based on the content and discussion, a PR can be opened. Lets not put the cart before the horses here and set expectation that a PR is required to raise a proposal in form of an RFC. 
If someone has some code, sure, they should open a PR, but here we say "a PR must be opened"....

Reply:
I am almost sure that sooner or later someone who opened the RFC issue would be asked to present it in the code. However we can leave opening PR as optional IMO and do not insist on RFC reporter.

Reply:
I would vote for enforcing the RFC in every case, just to be able to have a written record of the change in the form of GitHub issue. That will help with release management for example. One however may shorten the RFC contents (detailed description of the changes) if there is a PR accompanying it.